Avoid always building stack trace strings in c10::Error

IosifSpulber opened the pull request:
Summary:
The `c10::Error` class shows a lot of overhead building stack straces. Additionally, cost aside, this has caused capacity SEVs in the past since when something goes wrong: the exceptions thrown cause additional (and bigger) perf issues since they start putting together all the stack traces.

This is not needed the vast majority of time: an exception would be caught, there would be no logging (i.e., no call to `what()`) and the stack contents would be thrown away.

There are 3 issues here:
1) we have an abnormal number of exceptions thrown; the throw sites are suspicious, e.g., "submodule not found", or "model size mismatch"
2) symbolication is not as efficient as possible
3) we shouldn't care about the stack trace until we do

This diff tackles 3), although all the causes above should be looked at as they compound on each other. Intuitively, 3) should greatly alleviate the costs here.

The solution is borrowed from `FBException`: save the stack information, but the stringify-ing should be lazy:
* changed the `backtrace` expected by `c10::Error::Error` to a function that returns a backtrace instead
* the backtrace generator is invoked the first time `what()` (or, explicitly, `backtrace()`) is called; the result is afterward cached
* the macros constructing these exceptions generated the stacktrace with a cross-platform solution set; now, they build a function that returns a stacktrace. This requires snapshotting a `facebook::process::StackTrace`, and baking into the returned function. When called, this will actually invoke the `toString()` operation. Note that this is only available for the `FBCODE_CAFFE2` flavour - for the other platforms, we still compute the string right away.
* special care is needed for the hack of hijacking this stack trace generation and replacing it on-demand; namely, the introduced API needs to be compatible (and improve) the acrobatics in `SetStackTraceFetcher()` and `registerStacktraceFetcher()`.

Other things worth mentioning:
* for some reason `std::function` is 16-byte aligned, and the iOS flavour requires the exception class to be 8-byte aligned. There's a couple of solutions here, but I went for just wrapping the function in a `shared_ptr`.
* we can't use folly but for thread safety I went for `std::call_once`
* the added copy constructor / assignment operator are needed because the `std::once_flag` can't be copied and requires special handling
* although this class declares that it shouldn't be instantiated manually, there are some places that do so; they used a `""` backtrace, now changed to not supplying a backtrace generator at all in which case it defaults to a function returning an empty string backtrace.

Test Plan:
Added a bunch of relevant tests.

In particular, added tests to ensure the correct stack trace capture and proving that the stack trace is lazily generated.

Differential Revision: D54494101

Skylion007 commented on c10/util/Exception.cpp:
```diff
@@ -8,9 +8,21 @@
+  std::string msg,
+  std::optional<BacktraceGenerator> backtraceCallback,
+  const void* caller)
+    : msg_(std::move(msg)), caller_(caller) {
```
```suggestion
    : msg_(std::move(msg)), caller_(std::move(caller)) {
```

ezyang commented:
Most of the review is happening internally

IosifSpulber committed:
Make c10::Error empty backtrace as an optional argument (#122611)

Summary:

Split from the main diff in the stack.

Test Plan: Build validation should be enough.

Reviewed By: ezyang

Differential Revision: D55313410

IosifSpulber committed:
Avoid always building stack trace strings in c10::Error (#122086)

Summary:

The `c10::Error` class shows a lot of overhead building stack straces: [~1M $CPU-T1-equiv](https://fburl.com/strobelight/zbucahl6). Additionally, cost aside, this has caused capacity SEVs in the past since when something goes wrong: the exceptions thrown cause additional (and bigger) perf issues since they start putting together all the stack traces.

This is not needed the vast majority of time: an exception would be caught, there would be no logging (i.e., no call to `what()`) and the stack contents would be thrown away.

There are 3 issues here:
1) we have an abnormal number of exceptions thrown; the throw sites are suspicious, e.g., "submodule not found", or "model size mismatch"
2) symbolication is not as efficient as possible
3) we shouldn't care about the stack trace until we do

This diff tackles 3), although all the causes above should be looked at as they compound on each other. Intuitively, 3) should greatly alleviate the costs here.

The solution is borrowed from `FBException`: save the stack information, but the stringify-ing should be lazy:
* changed the `backtrace` expected by `c10::Error::Error` to a function that returns a backtrace instead
* the backtrace generator is invoked the first time `what()` (or, explicitly, `backtrace()`) is called; the result is afterward cached
* the macros constructing these exceptions generated the stacktrace with a cross-platform solution set; now, they build a function that returns a stacktrace. This requires snapshotting a `facebook::process::StackTrace`, and baking into the returned function. When called, this will actually invoke the `toString()` operation. Note that this is only available for the `FBCODE_CAFFE2` flavour - for the other platforms, we still compute the string right away.
* special care is needed for the hack of hijacking this stack trace generation and replacing it on-demand; namely, the introduced API needs to be compatible (and improve) the acrobatics in `SetStackTraceFetcher()` and `registerStacktraceFetcher()`.

Other things worth mentioning:
* for some reason `std::function` is 16-byte aligned, and the iOS flavour requires the exception class to be 8-byte aligned. There's a couple of solutions here, but I went for just wrapping the function in a `shared_ptr`.
* we can't use folly but for thread safety I went for `std::call_once`
* the added copy constructor / assignment operator are needed because the `std::once_flag` can't be copied and requires special handling
* although this class declares that it shouldn't be instantiated manually, there are some places that do so; they used a `""` backtrace, now changed to not supplying a backtrace generator at all in which case it defaults to a function returning an empty string backtrace.

Test Plan:
Added a bunch of relevant tests.

In particular, added tests to ensure the correct stack trace capture and proving that the stack trace is lazily generated.

Differential Revision: D54494101

swolchok commented:
was abandoned internally, closing PR
